app/src/main/java/com/otaman/notes/viewmodel/AllNotesViewModel.kt

```diff
@@ -5,15 +5,27 @@ import androidx.lifecycle.*
 import com.otaman.notes.model.Note
 import com.otaman.notes.model.NoteRepository
 import com.otaman.notes.model.room.NoteDatabase
+import com.otaman.notes.view.MainActivity
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
+import java.util.*
 
 class AllNotesViewModel(application: Application): AndroidViewModel(application) {
-    val dao = NoteDatabase.getDatabase(application).getNoteDao()
-    val repository: NoteRepository = NoteRepository(dao)
+    private val dao = NoteDatabase.getDatabase(application).getNoteDao()
+    private val repository: NoteRepository = NoteRepository(dao)
+    private val _searchResults = MutableLiveData<List<Note>>()
+    val searchResults: LiveData<List<Note>> = _searchResults
     val allNotes: LiveData<List<Note>> = repository.getAllNotes()
 
     fun deleteNote(note: Note) = viewModelScope.launch(Dispatchers.IO) {
         repository.deleteNote(note)
     }
+
+    fun searchNote(searchNote: String) = viewModelScope.launch(Dispatchers.IO) {
+        val result = allNotes.value?.filter { note ->
+            note.title.contains(searchNote, true) || note.content.contains(searchNote, true)
+        }.orEmpty()
+
+        _searchResults.postValue(result)
+    }

```

Comment:
It would be better to use `Dispatchers.Default` here

**IO Dispatcher:**
> It starts the coroutine in the IO thread, it is used to perform all the data operations such as networking, reading, or writing from the database, reading, or writing to the files eg: Fetching data from the database is an IO operation, which is done on the IO thread.

**Default Dispatcher:**

> It starts the coroutine in the Default Thread. We should choose this when we are planning to do Complex and long-running calculations, which can block the main thread and freeze the UI eg: Suppose we need to do the 10,000 calculations and we are doing all these calculations on the UI thread ie main thread, and if we wait for the result or 10,000 calculations, till that time our main thread would be blocked, and our UI will be frozen, leading to poor user experience. So in this case we need to use the Default Thread.

Reply:
Changed Dispatchers `.IO` thread to `.Default`